File: content/LLD/Reference-Code/1. OOPS Recap/1. Revision/movie_project.py
"""
Create a class Movie with the following:

Attributes:
movie_name —> name of the movie
total_seats —> total seats available in the theatre
ticket_price —> price per ticket
booked_seats —> starts at 0

Methods:
book_ticket(num_tickets) — books the given number of tickets. If enough seats are available,
confirm the booking and show the total amount to pay. If not,
show "Sorry, not enough seats available"

show_status() — displays movie name, seats available, and seats booked so far
"""


class Movie:
    def __init__(self, movie_name: str, total_seats: int, ticket_price: int):
        self.movie_name = movie_name
        self.total_seats = total_seats
        self.ticket_price = ticket_price
        self.booked_seats = 0

    def book_tickets(self, num_of_tickets: int):
        if num_of_tickets > self.total_seats - self.booked_seats:
            print("Sorry, not enough seats available")
        else:
            self.booked_seats += num_of_tickets
            self.total_seats -= num_of_tickets  # 30
            print(f"Your ticket is booked")
            print(f"Total price = {self.ticket_price * num_of_tickets}\n")

    def show_status(self) -> None:
        print(f"Movie name = {self.movie_name}")
        print(f"Seats available = {self.total_seats}")
        print(f"Total booked = {self.booked_seats}\n")


movie = Movie("Krish", 100, 499)
movie.show_status()
movie.book_tickets(70)

movie.show_status()

movie.book_tickets(70)

File: content/LLD/Reference-Code/1. OOPS Recap/1. Revision/MovieProject.java
/**
 * Create a class Movie with the following:
 *
 * Attributes:
 * movieName  -> name of the movie
 * totalSeats -> total seats available in the theatre
 * ticketPrice -> price per ticket
 * bookedSeats -> starts at 0
 *
 * Methods:
 * bookTickets(numOfTickets) -> books the given number of tickets. If enough seats are available,
 * confirm the booking and show the total amount to pay. If not,
 * show "Sorry, not enough seats available"
 *
 * showStatus() -> displays movie name, seats available, and seats booked so far
 */
class Movie {
    private String movieName;
    private int totalSeats;
    private int ticketPrice;
    private int bookedSeats;

    public Movie(String movieName, int totalSeats, int ticketPrice) {
        this.movieName = movieName;
        this.totalSeats = totalSeats;
        this.ticketPrice = ticketPrice;
        this.bookedSeats = 0;
    }

    public void bookTickets(int numOfTickets) {
        if (numOfTickets > this.totalSeats - this.bookedSeats) {
            System.out.println("Sorry, not enough seats available");
        } else {
            this.bookedSeats += numOfTickets;
            this.totalSeats -= numOfTickets;
            System.out.println("Your ticket is booked");
            System.out.println("Total price = " + (this.ticketPrice * numOfTickets) + "\n");
        }
    }

    public void showStatus() {
        System.out.println("Movie name = " + this.movieName);
        System.out.println("Seats available = " + this.totalSeats);
        System.out.println("Total booked = " + this.bookedSeats + "\n");
    }
}

public class MovieProject {
    public static void main(String[] args) {
        Movie movie = new Movie("Krish", 100, 499);
        movie.showStatus();
        movie.bookTickets(70);
        movie.showStatus();
        movie.bookTickets(70);
    }
}
